xds/internal/balancer/clusterresolver/e2e_test/balancer_test.go

```diff
@@ -302,3 +313,217 @@ func (s) TestErrorFromParentLB_ResourceNotFound(t *testing.T) {
+		BaseEjectionTime:         durationpb.New(30 * time.Second),
+		MaxEjectionTime:          durationpb.New(300 * time.Second),
+		MaxEjectionPercent:       wrapperspb.UInt32(10),
+		SuccessRateStdevFactor:   wrapperspb.UInt32(2000),
+		EnforcingSuccessRate:     wrapperspb.UInt32(50),
+		SuccessRateMinimumHosts:  wrapperspb.UInt32(10),
+		SuccessRateRequestVolume: wrapperspb.UInt32(50),
+	}
+	resources := e2e.UpdateOptions{
+		NodeID:         nodeID,
+		Clusters:       []*v3clusterpb.Cluster{cluster},
+		Endpoints:      []*v3endpointpb.ClusterLoadAssignment{e2e.DefaultEndpoint(edsServiceName, "localhost", []uint32{testutils.ParsePort(t, server.Address)})},
+		SkipValidation: true,
+	}
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+	if err := managementServer.Update(ctx, resources); err != nil {
+		t.Fatal(err)
+	}
+
+	// Create xDS client, configure cds_experimental LB policy with a manual
+	// resolver, and dial the test backends.
+	_, cleanup = setupAndDial(t, bootstrapContents)
+	defer cleanup()
+
+	// The priority configuration generated should have Outlier Detection as a
+	// direct child due to Outlier Detection being turned on.
+	wantCfg := &priority.LBConfig{
+		Children: map[string]*priority.Child{
+			"priority-0-0": {
+				Config: &iserviceconfig.BalancerConfig{
+					Name: outlierdetection.Name,
+					Config: &outlierdetection.LBConfig{
+						Interval:           iserviceconfig.Duration(10 * time.Second), // default interval
+						BaseEjectionTime:   iserviceconfig.Duration(30 * time.Second),
+						MaxEjectionTime:    iserviceconfig.Duration(300 * time.Second),
+						MaxEjectionPercent: 10,
+						SuccessRateEjection: &outlierdetection.SuccessRateEjection{
+							StdevFactor:           2000,
+							EnforcementPercentage: 50,
+							MinimumHosts:          10,
+							RequestVolume:         50,
+						},
+						ChildPolicy: &iserviceconfig.BalancerConfig{
+							Name: clusterimpl.Name,
+							Config: &clusterimpl.LBConfig{
+								Cluster:        clusterName,
+								EDSServiceName: edsServiceName,
+								ChildPolicy: &iserviceconfig.BalancerConfig{
+									Name: wrrlocality.Name,
+									Config: &wrrlocality.LBConfig{
+										ChildPolicy: &iserviceconfig.BalancerConfig{
+											Name: roundrobin.Name,
+										},
+									},
+								},
+							},
+						},
+					},
+				},
+				IgnoreReresolutionRequests: true,
+			},
+		},
+		Priorities: []string{"priority-0-0"},
+	}
+
+	select {
+	case jsonCfg := <-testChildPolicy.cfgCh:
+		gotCfg := &priority.LBConfig{}
+		if err := json.Unmarshal(jsonCfg, gotCfg); err != nil {
+			t.Fatalf("Failed to unmarshal LB config received by child policy: %v", err)
+		}
+		if diff := cmp.Diff(wantCfg, gotCfg); diff != "" {
+			t.Fatalf("Child policy received unexpected diff in config (-want +got):\n%s", diff)
+		}
+	case <-ctx.Done():
+		t.Fatalf("Timeout when waiting for child policy to receive its configuration")
+	}
+}
```

Comment:
Unlike the test above, this check/recv is not in a goroutine. Is there a reason for the divergence? (is it this is balancer.Balancer API vs. balancer.ClientConn in test above) Also, similar question here, I don't get how this isn't racy. Are we guaranteed this select is running by the time we send on cfgCh on line 348? Or can that send come before this is running losing this check and making test flaky?

Reply:
I don't need a goroutine here because I'm not performing an RPC here. Technically, we don't need to perform an RPC in the previous test as well (to get the subChannel to READY), but I thought it was more appropriate to do the RPC there, because we also want to ensure that a READY subchannel also means successful RPCs.

This is not racy here because we expect only one config update to be pushed to the child policy, and the channel used by the wrapped balancer has a buffer of 1 (I thought I had a buffer of 1, but didn't have it, so I added it now). This means that the writer will not be blocked even if the reading goroutine has not started running.

Reply:
I meant racy with the send being lost do the default in the case of the channel not having buffer of 1. Now that is has buffer of 1 and we expect 1 config update, this check sgtm. Either it sends before or after it gets to this select, and either case works :).